Add asymmetric key generation/import/export + refactor FFI helper amenities

Xanewok opened the pull request:
This... is rather a long list and more of a dump rather than a specific pull request that I'd expect to be merged.

Everything related to dynamic structures between https://github.com/emgre/win-crypto-ng/commit/1bec3043d749ba63f1e7eeda4bda82413ff92067 and https://github.com/emgre/win-crypto-ng/commit/9bbf91e47a00ad5e8eb4d1569aca04ba53b2de45 can more or less be simply squashed to a single commit implementing blob support and supporting byte conversions.

The history is long because I experimented with 3 different approaches to blob (basically struct with header and some extra bytes) types:
- Defining different traits for generic `AsBytes` (TypedBlob) implementee used for dynamic field access 
  1. TypedBlob was IMO a too heavy tool
  2. Needs to import a trait to access the dynamic fields
- custom Rust DSTs with a very generic (Header, Tail) type pair + trait used to reinterpret the Tail data 
  1. it infected 95% of signatures that touched the DST type
  2. overly generic and ironically prevented from using generic default impls for some tasks
  3. still needed to call a `view()` returning struct with refs to dynamic fields (all of them had to be calculated)
- custom Rust DSTs with `[u8]` tail, inherent impls for dynamic field getters
  1. no lifetime pollution
  2. field getters are inherent so no need to import or define traits
  3. unfortunately can't be implemented outside of the crate due to inherent impls but could be tweaked to be able to define blobs outside of the crate
  4. Relies on more or less well-defined casts to slice pointers, which is better than just `transmuting` a reference (like in `TypedBlob`) to a sized header (even if we're checking the invariants) and expecting the unsized allocation to kind of be there

Omitted some details for brevity, can go more in-depth if needed.

After that dyn struct commit series there's some reorganizings and minor improvements to the remaining code and thus can be omitted for now.

@emgre thoughts?

EDIT: ~Also please bear with me, this is not formatted yet but hopefully clippy-happy. All tests pass locally~

Closes #15
Closes #16
Closes #19

Xanewok committed:
Move key handle wrapper to a separate module

Xanewok committed:
Add keypair exporting

Xanewok committed:
Add a way to export different asymmetric key kinds

Xanewok committed:
Add TypedBlob::as_bytes

This is a convenience function that can be later used to dynamically
access fields for a dynamic structs.

Xanewok committed:
Implement Deref<Target = T> for TypedBlob<T>

After all, this is more of a smart pointer facility rather than anything
else - we shouldn't get through the hassle of repeatedly doing .as_ref()
only to access relevant fields or methods.

Xanewok committed:
Implement AsRef for TypedBlob in terms of the inner AsRef target instead

This allows us to cut through a layer of abstraction, where the inner
type is possibly a newtype wrapper on its own.

While technically this doesn't make the code more expressive (we
could've used `blob.deref().as_ref()` directly), it does allow us to
express more generic trait bounds, e.g. `AsRef<BCRYPT_DH_KEY_BLOB>` for
some `TypedBlob<DhWrapper>` that may carry additional compile-time
information.

Xanewok committed:
Use newtypes to hold typed key blobs

This allows us to further implement extensions to the values holding the
keys themselves, such as dynamically computed fields that are not
available via the original struct definition (it's only header data)

Xanewok committed:
Implement Debug for TypedBlob

Xanewok committed:
Introduce dyn_struct! macro

Xanewok committed:
Add a note that we only support slices in TypedBlob

Xanewok committed:
Add asymmetric crypto stub

Xanewok committed:
Add asymmetric crypto key pairs

Xanewok committed:
Add more concrete/high-level asymmetric key types

Xanewok committed:
Add type-safe key generation

Xanewok committed:
Support generic ECC curves

Xanewok committed:
Redesign dyn_struct contraption

Xanewok committed:
Make local Result type more flexible

Xanewok committed:
Improve safety in byte conversions

Xanewok committed:
Safely handle owned [u16] in WindowsString constructor

Xanewok committed:
Remove unused types

Xanewok committed:
Implement from/into byte conversion, e.g. for custom DSTs

Xanewok committed:
Rename DynStruct to Blob

Xanewok committed:
Prefer {As,From}Bytes Blob impl over manual impls

Xanewok committed:
Rename WindowsString to WideCString

To be more in line with convention of owned `(C)String` types

Xanewok committed:
Remove unused set_property_dyn

Xanewok committed:
impl From<&str> for WideCString

Xanewok committed:
Ensure that get_property calls write to entire allocation

Xanewok committed:
Split impls in helpers.rs into relevant submodules

- Handle-related keys are now in handle.rs
- Property setting/getting is now in src/property.rs

Xanewok committed:
Remove unused list_algorithms in helpers.rs

Xanewok committed:
Add missing select modules' documentation

Xanewok committed:
Move KeyHandle to handle module

Xanewok committed:
Downgrade MSRV to at least 1.37 again

Xanewok committed:
Make curves public in Ecdh/Ecdsa algo ZSTs

emgre commented on src/lib.rs:
```diff
@@ -4,13 +4,16 @@ use winapi::shared::ntstatus;
 
+pub mod asymmetric;
 pub mod buffer;
+pub mod handle;
```
This does not need to be public. The user should always interact with a `SymmetricAlgorithmKey` or a `AsymmetricKey`. Also, it could be move with the helpers.

Xanewok replied:
I'm not sure helpers is a good idea, since it mostly collects FFI-specific code like inline blobs or window-specific strings. For example, at some point we might even replace WideCString with a the `widestring` crate or from_bytes/as_bytes with the `bytemuck` crate, whereas handle module is only related to CNG.

I'll mark this module as private.

emgre commented on src/property.rs:
```diff
@@ -102,12 +190,23 @@ impl Property for KeyLengths {
+impl Property for DhParameters {
+    const IDENTIFIER: &'static str = BCRYPT_DH_PARAMETERS;
+    type Value = [u8];
+}
```
`DsaParameters` and `DhParameters` needs documentation.

emgre commented on src/lib.rs:
```diff
@@ -4,13 +4,16 @@ use winapi::shared::ntstatus;
 pub mod symmetric;
 
-mod helpers;
+pub mod helpers;
```
This doesn't need to be all public.

Xanewok replied:
This exports `Blob`, `BlobLayout` and the `From/AsBytes` (which are part of `BlobLayout`) that are part of the public API. I believe we should leave that until we potentially change the public API not to touch blob types at all.

emgre commented on src/helpers/bytes.rs:
```diff
@@ -0,0 +1,181 @@
+//! Byte fiddling utilities.
+//!
+//! Define [FromBytes] and [AsBytes] traits, which allow for safe data conversion
```
The links doesn't work.

emgre commented on src/helpers/string.rs:
```diff
@@ -0,0 +1,54 @@
+
+use winapi::shared::ntdef::LPCWSTR;
+
+pub struct WideCString {
```
We should add a few unit tests of `WideCString`, just to sure.

emgre commented on src/helpers/bytes.rs:
```diff
@@ -0,0 +1,181 @@
+
+use winapi::shared::bcrypt;
+unsafe impl Pod for u32 {} // Ignores endianness
+unsafe impl Pod for bcrypt::BCRYPT_KEY_LENGTHS_STRUCT {}
```
The bcrypt `Pod` impl could be moved to `key.rs`. Also, we should import `bcrypt::*` and remove the `bcrypt::` to be consistent.

emgre commented on src/helpers/blob.rs:
```diff
@@ -0,0 +1,240 @@
+/// but with trailing data of size dependent on the header field values.
+///
+/// # Layout
+/// The structure is marked as `#[repr(packed)]` to be layout-compatible with
```
It's actually `#[repr(C,packed)]`

emgre commented on src/helpers/blob.rs:
```diff
@@ -0,0 +1,240 @@
+        // `Self::from_boxed`, which requires that the source reference is
+        // aligned at least as `T::Header` and since `Blob` is
+        // `#[repr(C)]` (and so is `T::Header`, because it  implements `AsBytes`
+        // which requires which requires being `#[repr(C)]`), so the reference
```
which requires which requires

emgre commented on src/helpers/blob.rs:
```diff
@@ -0,0 +1,240 @@
+                let align = std::mem::align_of_val(header);
+                let tail_padding = (align - (tail_len % align)) % align;
+
+                dbg!(header_len);
```
These `dbg!` should be removed.

Xanewok replied:
Yep, my bad!

emgre commented on src/helpers/blob.rs:
```diff
@@ -0,0 +1,240 @@
+/// layout.
+/// Assumes a contiguous byte buffer.
+#[macro_export]
+macro_rules! blob {
```
This is very clever. The macro definition is not so easy to read, but its usage is really nice. I like it!

emgre commented on src/asymmetric/ecc.rs:
```diff
@@ -0,0 +1,77 @@
+}
+
+ // Required due to 1.37 being MSRV
+ // TODO: Replace with #[non_exhaustive] once we bump MSRV to 1.40
```
The `non_exhaustive` is temporary, right? Once we have all named curves implementend, it doesn't make sense to have it non-exhaustive.

Xanewok replied:
It was temporary but then again it's possible for CNG to add support for more curves as time goes on. We could drop it for now and revisit once something like that actually happens :)

Xanewok replied:
Removed the `non_exhaustive` attribute for now.

emgre commented on Cargo.toml:
```diff
@@ -15,7 +15,7 @@ license = "BSD-3-Clause"
 
 [dependencies]
-winapi = { version = "0.3", features = ["bcrypt", "ntstatus"] }
+winapi = { version = "0.3", features = ["bcrypt", "ntstatus", "winerror", "winbase", "debug"] }
```
Is `debug` really required?

Xanewok replied:
It's for `Debug` impl for winapi structs, IIRC I needed that when I was debugging something but we definitely not should unconditionally enable it

emgre reviewed (commented):
- I like the `Erased` type as well as the typed values. Makes it very versatile.
- For modules, I prefer to use `mod.rs` inside a directory instead of a seperate file with the same name as the folder. I saw that `asymmetric.rs` and `helpers.rs` were like that.
- Regarding `bytes.rs`, I kinda have to rely on you that it works. I understand the code but my knowledge of alignement requirements and such is limited.
- I still have to look at the `asymmetric` module in more details. From a first quick look, the public API seems a bit large.

Xanewok committed:
Make handle module private

Xanewok committed:
Document DH/DSA parameter properties

Xanewok committed:
Polish and test WideCString

Xanewok committed:
Move Pod impls to relevant modules

Xanewok committed:
Tweak documentation

Xanewok committed:
Use AsBytes rather than directly slicing blob header

Xanewok committed:
Unify key blob payload struct names

Xanewok committed:
Add missing documentation for key blob types

Xanewok committed:
Make export_blobs syntax symmetrical to the import one

Xanewok commented:
Squashed the noisiest commits, rewrote the history to fix the LF endings (we should probably include an .editorconfig with LF setting) and formatting, polished some more documentation.

The only things left are some minor clippy lints (one safety docs and some by-ref/value bits) and hopefully this should be good to go!

Xanewok committed:
Document unsafety in FromBytes::min_layout

Xanewok committed:
Appease clippy::trivially_copy_pass_by_ref

Xanewok committed:
Silence false positive wrong_self_convention lint

Xanewok commented:
This should be now ready for final review 🥳

Xanewok committed:
Move submodule roots to mod.rs

Xanewok commented:
> For modules, I prefer to use mod.rs inside a directory

Updated. I personally prefer named approach because it's easier to navigate opened tabs in the editor but I can see how it makes more sense in terms of file layout 😃

Xanewok committed:
Note that DSA V2 blob applies to *above* 1024 bit keys

Xanewok committed:
Use actual DSA blob v2 layout

Xanewok committed:
Silence yet another Clippy warning

This is a nightly-only feature - something we can't enable because not
only do we build on stable, we require MSRV of 1.37.

Xanewok commented:
Closing in favor of #27.
